substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -33,16 +33,207 @@ use codec::{Compact, Decode, Encode, EncodeLike, Error, Input};
 use core::fmt;
 use scale_info::{build::Fields, meta_type, Path, StaticTypeInfo, Type, TypeInfo, TypeParameter};
 use sp_io::hashing::blake2_256;
+use sp_weights::Weight;
+
+/// Type to represent the version of the [Extension](TransactionExtension) used in this extrinsic.
+pub type ExtensionVersion = u8;
+/// Type to represent the extrinsic format version which defines an [UncheckedExtrinsic].
+pub type ExtrinsicVersion = u8;
 
 /// Current version of the [`UncheckedExtrinsic`] encoded format.
 ///
 /// This version needs to be bumped if the encoded representation changes.
 /// It ensures that if the representation is changed and the format is not known,
 /// the decoding fails.
-const EXTRINSIC_FORMAT_VERSION: u8 = 4;
+pub const EXTRINSIC_FORMAT_VERSION: ExtrinsicVersion = 5;
+/// Current version of the [Extension](TransactionExtension) used in this
+/// [extrinsic](UncheckedExtrinsic).
+///
+/// This version needs to be bumped if there are breaking changes to the extension used in the
+/// [UncheckedExtrinsic] implementation.
+const EXTENSION_VERSION: ExtensionVersion = 0;
 
 /// The `SignaturePayload` of `UncheckedExtrinsic`.
-type UncheckedSignaturePayload<Address, Signature, Extra> = (Address, Signature, Extra);
+pub type UncheckedSignaturePayload<Address, Signature, Extension> = (Address, Signature, Extension);
+
+impl<Address: TypeInfo, Signature: TypeInfo, Extension: TypeInfo> SignaturePayload
+	for UncheckedSignaturePayload<Address, Signature, Extension>
+{
+	type SignatureAddress = Address;
+	type Signature = Signature;
+	type SignatureExtra = Extension;
+}
+
+/// A "header" for extrinsics leading up to the call itself. Determines the type of extrinsic and
+/// holds any necessary specialized data.
+#[derive(Eq, PartialEq, Clone)]
+pub enum Preamble<Address, Signature, Extension> {
+	/// An extrinsic without a signature or any extension. This means it's either an inherent or
+	/// an old-school "Unsigned" (we don't use that terminology any more since it's confusable with
+	/// the general transaction which is without a signature but does have an extension).
+	///
+	/// NOTE: In the future, once we remove `ValidateUnsigned`, this will only serve Inherent
+	/// extrinsics and thus can be renamed to `Inherent`.
+	Bare(ExtrinsicVersion),
+	/// An old-school transaction extrinsic which includes a signature of some hard-coded crypto.
+	Signed(Address, Signature, ExtensionVersion, Extension, ExtrinsicVersion),
```

Comment:
Is there any reason why `ExtrinsicVersion` is added here? As far as I understood, the shape of an Extrinsic of any kind is `(ExtrinsicFormat, ExtrinsicPayload)`, where extrinsic format already includes the `ExtrinsicVersion` (`0xTTVV_VVVV`, where `TT` defines Bare, Signed, or General and `VV_VVVV` defines the version).

Reply:
I think this is only a rust representation of the preamble. You are correct both the variant and the extrinsic version is encoded in a single byte, the extrinsic format.

Reply:
But just because it is handy there's an extra byte being added to the extrinsic?
I'd revise it, it is certainly weird to repeat it. A byte per transaction is certainly something.

Reply:
no it is only in the in-memory representation, not encoded representation. When the preamble is encoded it goes back to one byte for extrinsic format.

The impact of one byte here is negligible.

Reply:
If the encoded version does not contain it then we're good to go 👍🏻